File: src/application/flashcard_session.py
import random
from datetime import datetime

from domain.models import normalize_words_collection


class FlashcardSession:
    def __init__(self, words, rng=None):
        self.words = normalize_words_collection(words)
        self.rng = rng or random.Random()
        self.current_word = None
        self.current_context = None
        self.last_answer_known = None
        self.last_context_ids = {}

    def next_card(self):
        if not self.words:
            self.current_word = None
            self.current_context = None
            self.last_answer_known = None
            return None

        candidates = self.words
        if self.current_word and len(self.words) > 1:
            candidates = [word for word in self.words if word is not self.current_word]

        self.current_word = self.rng.choice(candidates)
        self.current_context = self._select_context(self.current_word)
        self.last_answer_known = None

        return {
            "word": self.current_word,
            "context": self.current_context,
        }

    def answer_card(self, known):
        if not self.current_word:
            return None

        if known:
            self.current_word.register_correct(now=datetime.now())
        else:
            self.current_word.register_wrong()

        self.last_answer_known = known
        return {
            "known": known,
            "word": self.current_word,
            "context": self.current_context,
        }

    def current_stats(self):
        if not self.current_word:
            return {"correct": 0, "wrong": 0, "last_correct": False}

        score = self.current_word.vocabulary_score()
        score["last_correct"] = self.current_word.get("last_correct", False)
        return score

    def _select_context(self, word):
        contexts = word.get("_contexts", [])
        if not contexts:
            return None

        if len(contexts) == 1:
            context = contexts[0]
        else:
            previous_context_id = self.last_context_ids.get(word.get("_word_id"))
            candidates = [
                context
                for context in contexts
                if context.get("id") != previous_context_id
            ]
            context = self.rng.choice(candidates or contexts)

        context_id = context.get("id")
        if context_id:
            self.last_context_ids[word.get("_word_id")] = context_id

        return context

File: flutter_app/lib/services/flashcard_session.dart
import 'dart:math';

import '../models/learning_context.dart';
import '../models/learning_word.dart';

enum FlashcardDeckMode {
  allWords,
  withContexts,
  needsReview,
}

class FlashcardCard {
  const FlashcardCard({
    required this.word,
    required this.context,
  });

  final LearningWord word;
  final LearningContext? context;
}

class FlashcardAnswerResult {
  const FlashcardAnswerResult({
    required this.known,
    required this.word,
    required this.context,
  });

  final bool known;
  final LearningWord word;
  final LearningContext? context;
}

typedef WordProgressCallback = void Function(LearningWord word);

class FlashcardStats {
  const FlashcardStats({
    required this.correct,
    required this.wrong,
    required this.total,
    required this.lastCorrect,
  });

  final int correct;
  final int wrong;
  final int total;
  final String? lastCorrect;
}

class FlashcardSession {
  FlashcardSession(
    List<LearningWord> words, {
    Random? rng,
    this.deckMode = FlashcardDeckMode.allWords,
  })  : _sourceWords = List<LearningWord>.from(words),
        _rng = rng ?? Random() {
    _rebuildDeck();
  }

  final Random _rng;
  final Map<String, String> _lastContextIds = <String, String>{};
  final List<LearningWord> _sourceWords;
  final Set<String> _seenWordIds = <String>{};
  List<LearningWord> _words = <LearningWord>[];

  FlashcardDeckMode deckMode;

  LearningWord? currentWord;
  LearningContext? currentContext;
  bool? lastAnswerKnown;
  int answeredCount = 0;

  int get wordCount => _words.length;
  int get seenCount => _seenWordIds.length;
  double get sessionProgress => wordCount == 0 ? 0 : seenCount / wordCount;

  int get wordsWithContextsCount {
    return _words.where((word) => word.contexts.isNotEmpty).length;
  }

  void setDeckMode(FlashcardDeckMode mode) {
    if (deckMode == mode) {
      return;
    }

    deckMode = mode;
    _rebuildDeck();
  }

  FlashcardCard? nextCard() {
    if (_words.isEmpty) {
      currentWord = null;
      currentContext = null;
      lastAnswerKnown = null;
      return null;
    }

    var candidates = _words;
    if (currentWord != null && _words.length > 1) {
      candidates = _words
          .where((word) => word.wordId != currentWord!.wordId)
          .toList(growable: false);
    }

    currentWord = candidates[_rng.nextInt(candidates.length)];
    currentContext = _selectContext(currentWord!);
    lastAnswerKnown = null;
    _seenWordIds.add(currentWord!.wordId);

    return FlashcardCard(
      word: currentWord!,
      context: currentContext,
    );
  }

  FlashcardAnswerResult? answerCard(bool known) {
    final activeWord = currentWord;
    if (activeWord == null) {
      return null;
    }

    final updatedWord = activeWord.copyWith(
      correct: known ? activeWord.correct + 1 : activeWord.correct,
      wrong: known ? activeWord.wrong : activeWord.wrong + 1,
      lastCorrect: known ? DateTime.now().toIso8601String() : activeWord.lastCorrect,
    );

    final index = _words.indexWhere((word) => word.wordId == activeWord.wordId);
    if (index != -1) {
      _words[index] = updatedWord;
    }

    currentWord = updatedWord;
    lastAnswerKnown = known;
    answeredCount += 1;

    return FlashcardAnswerResult(
      known: known,
      word: updatedWord,
      context: currentContext,
    );
  }

  FlashcardStats currentWordStats() {
    final activeWord = currentWord;
    if (activeWord == null) {
      return const FlashcardStats(
        correct: 0,
        wrong: 0,
        total: 0,
        lastCorrect: null,
      );
    }

    return FlashcardStats(
      correct: activeWord.correct,
      wrong: activeWord.wrong,
      total: activeWord.correct + activeWord.wrong,
      lastCorrect: activeWord.lastCorrect,
    );
  }

  LearningContext? _selectContext(LearningWord word) {
    final contexts = word.contexts;
    if (contexts.isEmpty) {
      return null;
    }

    late final LearningContext context;
    if (contexts.length == 1) {
      context = contexts.first;
    } else {
      final previousContextId = _lastContextIds[word.wordId];
      final candidates = contexts
          .where((entry) => entry.contextId != previousContextId)
          .toList(growable: false);
      final pool = candidates.isEmpty ? contexts : candidates;
      context = pool[_rng.nextInt(pool.length)];
    }

    if (context.contextId.isNotEmpty) {
      _lastContextIds[word.wordId] = context.contextId;
    }

    return context;
  }

  void _rebuildDeck() {
    _words = _buildDeck(_sourceWords, deckMode);
    currentWord = null;
    currentContext = null;
    lastAnswerKnown = null;
    answeredCount = 0;
    _lastContextIds.clear();
    _seenWordIds.clear();
  }

  List<LearningWord> _buildDeck(
    List<LearningWord> words,
    FlashcardDeckMode mode,
  ) {
    switch (mode) {
      case FlashcardDeckMode.allWords:
        return List<LearningWord>.from(words);
      case FlashcardDeckMode.withContexts:
        return words
            .where((word) => word.contexts.isNotEmpty)
            .toList(growable: false);
      case FlashcardDeckMode.needsReview:
        final reviewWords = words
            .where((word) => word.wrong > 0 || word.wrong > word.correct)
            .toList(growable: false);
        return reviewWords;
    }
  }
}
